content/item-hash/index.md

```diff
@@ -0,0 +1,257 @@
+
+
+### Redaction
+
+Redaction works in the same way as
+[objecthash](https://github.com/benlaurie/objecthash/blob/d1e3d6079fc16f8f542183fb5b2fdc11d9f00866/README.md#L65).
+
+In summary, these two items are equivalent:
+
+```elm
+i_0 =
+  Dict [ ("foo", "abc")
+       , ("bar", "xyz")
+       ]
+
+i_1 =
+  Dict [ ("foo", "**REDACTED**2a42a9c91b74c0032f6b8000a2c9c5bcca5bb298f004e8eff533811004dea511")
+       , ("bar", "xyz")
+       ]
+```
+
+The reason they are equivalent is because the hash for `"abc"` is
+"2a42a9c91b74c0032f6b8000a2c9c5bcca5bb298f004e8eff533811004dea511". So when
+the hashing algorithm is applied to the whole item, the redacted value is used
+as is (without the redaction tag).
+
+### Example
+
+Note: To simplify the code, the hashing algorithm `SHA-256` is implied. A full
+implementation must acknowledge the possibility of a different algorithm.
+
+To walk through the algorithm I'll use the following item:
+
+```elm
+Dict
+  [ ("id", "GB")
+  , ("official-name", "The United Kingdom of Great Britain and Northern Ireland")
+  , ("name", "United Kingdom")
+  , ("citizen-names", Set ["Briton", "British citizen"])
+  ]
+```
+
+Which is serialised as JSON as:
+
+```json
+{
+  "id": "GB",
+  "official-name": "The United Kingdom of Great Britain and Northern Ireland",
+  "name": "United Kingdom",
+  "citizen-names": ["Briton", "British citizen"]
+}
+```
+
+The following definitions set the context of operation:
+
+```elm
+type Tag
+  = Dict
+  | Set
+  | String
+
+tagToChar : Tag -> Char
+tagToChar tag =
+  case tag of
+    Dict ->
+      'd'
+    Set ->
+      's'
+    String ->
+      'u'
+```
+
+```elm
+hashTagged : String -> Hash
+
+hashString : String -> Hash
+
+hashSet : Set -> Hash
+
+hashValue : Value -> Hash
+
+hashPair : (String, Value) -> List Hash
+```
+
+```elm
+hash : Item -> Hash
+```
+
+The `hash` function, step by step would roughly look like:
+
+```elm
+id = hashPair ("id", "GB")
+id == "17b788a70eeccbdc2fcb2d2d3db216c02fa88ac668beeb164bb2328c864bf3f4fff7021c7df4426be0f9a3c83f236eb6f85d159e624b010d65e6dde267889c21"
+
+officialName = hashPair ("official-name", "The United Kingdom of Great Britain and Northern Ireland")
+officialName == "cf09bea8c0107bd2150b073150d48db0a5b24c83defc7960ed698378d9f84b93bf1860175c77869938cf9f4b37edb00f2f387be7b361f9c2c4a2ac202c1ba2e5"
+
+name = hashPair ("name", "United Kingdom")
+name == "5c0be87ed7434d69005f8bbd84cad8ae6abfd49121b4aaeeb4c1f4a2e298771194099b1e0b9a1e673bafee513080197fa1980895ca27e091fdd4c54fab2bed24"
+
+citizenNames = hashPair ("citizen-names", Set ["Briton", "British citizen"])
+citizenNames == "bb3a7ac86d4f90c20d099992de0bd09bf3c4f27169c2cd873836762b01d5a2be16897987a6ee59d9ffdb456ed02df34a79b05346498d4360172568101ae157c1"
+```
+
+Then combine the partial results and tag it as a dictionary:
+
+```elm
+itemHash = hashDict [id, officialName, name, citizenNames]
+
+itemHash == "45d9392ad17cead3fa46501eba3e5ac237cb46a39f1e175905f00ef6a6667257"
```

Comment:
I can't get this value. It would be good to show you what I'm doing to see where I'm going wrong.

Reply:
I got this after trying again so ignore this comment!